[Messenger] Add HandlerArgumentsStamp

enumag opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #31075
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/17174

As discussed in #31075 sometimes it's desirable for the messenger handler to receive additional argument than just the message itself. I understand the voiced concerns about passing the entire envelope but instead of that we could use the approach from this PR which doesn't add any additional arguments by default but is actually even more powerful since it gives the user full control what should be sent to the handler if desired.

This is just a prototype of course. With https://github.com/symfony/symfony/pull/45377 in mind I used a readonly property from PHP 8.1 but of course such details can be easily adjusted.

Let me know if such feature is wanted in Symfony. If yes then I'll add tests and a doc PR.

enumag commented:
Can I get some feedback on this feature?

fabpot commented on src/Symfony/Component/Messenger/Stamp/HandlerArgumentsStamp.php:
```diff
@@ -0,0 +1,10 @@
+<?php declare(strict_types = 1);
```
`declare` should be removed (as we are using that in Symfony).

fabpot commented on src/Symfony/Component/Messenger/Middleware/HandleMessageMiddleware.php:
```diff
@@ -143,4 +144,21 @@ private function messageHasAlreadyBeenHandled(Envelope $envelope, HandlerDescrip
+        object $message,
+        ?Acknowledger $ack,
+        ?HandlerArgumentsStamp $handlerArgumentsStamp,
+    ): mixed {
```
Should be on one line.

fabpot reviewed (changes requested):
LGTM

enumag commented:
@fabpot Thanks for the review. I changed the code according to your comments and added a test.

The CI failure seems unrelated.

Status: Needs review

fabpot commented on src/Symfony/Component/Messenger/Stamp/HandlerArgumentsStamp.php:
```diff
@@ -0,0 +1,23 @@
+final class HandlerArgumentsStamp implements NonSendableStampInterface
+{
+    public function __construct(
+        public readonly array $additionalArguments,
```
As we are not using public properties in Symfony yet, I think we should make it private and create a proper getter (for consistency with the rest of the code base).

fabpot reviewed (approved):
Just a small code change to make it more consistent with the rest of Symfony.

enumag commented:
@fabpot Updated. Is it okay to keep the promoted property or should I use the non-promoted syntax?

enumag committed:
Add HandlerArgumentsStamp

enumag commented:
Tests are failing but once again I don't think it's related to my code.

fabpot commented:
Thank you @enumag.

melkamar commented:
What is the reasoning behind making the stamp implement a `NonSendableStampInterface`? I needed to customize the arguments for a handler reading this from the transport and it seems like I cannot. Is there a technical reason why not?
